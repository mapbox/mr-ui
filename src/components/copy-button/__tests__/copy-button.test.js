import React from 'react';
import { shallow } from 'enzyme';
import delay from 'delay';
import { testCases } from './copy-button-test-cases';
import getWindow from '../../utils/get-window';
import Clipboard from 'clipboard/dist/clipboard.min.js';

const mockUserAgent =
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/64.0.3282.186 Safari/537.36';

jest.mock('clipboard/dist/clipboard.min.js');

Clipboard.isSupported = () => true;

jest.mock('../../utils/get-window', () => {
  return jest.fn();
});

describe('CopyButton', () => {
  let testCase;
  let wrapper;

  beforeEach(() => {
    getWindow.mockReturnValue({
      navigator: {
        userAgent: mockUserAgent
      }
    });
  });

  describe(testCases.defaults.description, () => {
    beforeEach(() => {
      testCase = testCases.defaults;
      wrapper = shallow(
        React.createElement(testCase.component, testCase.props)
      );
    });

    test('renders as expected', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('changes state on click', () => {
      wrapper.find('button').prop('onClick')();
      wrapper.update();
      expect(wrapper).toMatchSnapshot();
      return delay(820).then(() => {
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
      });
    });

    test('focusTrapPaused false/undefined updates container element when setting clipboard', () => {
      wrapper.instance().setClipboard('cool container');

      expect(Clipboard).toHaveBeenCalledWith('cool container', {
        container: 'cool container'
      });
    });
  });

  describe(testCases.allProps.description, () => {
    beforeEach(() => {
      testCase = testCases.allProps;
      wrapper = shallow(
        React.createElement(testCase.component, testCase.props)
      );
    });

    test('renders as expected', () => {
      expect(wrapper).toMatchSnapshot();
    });

    // Extra classes need to carry over to both states
    test('changes state on click', () => {
      wrapper.find('button').prop('onClick')();
      wrapper.update();
      expect(wrapper).toMatchSnapshot();
      return delay(820).then(() => {
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
      });
    });

    test('calls onCopy callback', () => {
      wrapper.find('button').prop('onClick')();
      expect(testCase.props.onCopy).toHaveBeenCalledTimes(1);
      expect(testCase.props.onCopy).toHaveBeenCalledWith(testCase.props.text);
    });

    test('focusTrapPaused true does not update container element when setting clipboard', () => {
      wrapper.instance().setClipboard('cool container');

      expect(Clipboard).toHaveBeenCalledWith('cool container', {});
    });
  });
});
