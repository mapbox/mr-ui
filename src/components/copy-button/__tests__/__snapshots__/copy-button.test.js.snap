// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyButton Defaults changes state on click 1`] = `
<Tooltip
  alignment="center"
  block={false}
  coloring="light"
  content="Copy"
  disabled={true}
  maxWidth="medium"
  padding="small"
  placement="top"
  respondsToClick={false}
  textSize="s"
>
  <button
    className="btn btn--xs py3 px3 round"
    data-clipboard-text="Copied by the default test case"
    onClick={[Function]}
    type="button"
  >
    <Icon
      inline={false}
      name="check"
      passthroughProps={Object {}}
      size={18}
    />
    <Popover
      accessibleTitle={null}
      alignment="center"
      allowPlacementAxisChange={true}
      clickOutsideCloses={true}
      coloring="light"
      containWithinViewport={true}
      escapeCloses={true}
      getAnchorElement={[Function]}
      hasPointer={true}
      hideWhenAnchorIsOffscreen={true}
      padding="small"
      placement="top"
      receiveFocus={true}
      trapFocus={false}
      zIndex={1}
    >
      <div
        className="txt-s"
      >
        Copied!
      </div>
    </Popover>
  </button>
</Tooltip>
`;

exports[`CopyButton Defaults changes state on click 2`] = `
<Tooltip
  alignment="center"
  block={false}
  coloring="light"
  content="Copy"
  disabled={false}
  maxWidth="medium"
  padding="small"
  placement="top"
  respondsToClick={false}
  textSize="s"
>
  <button
    className="btn btn--xs py3 px3 round"
    data-clipboard-text="Copied by the default test case"
    onClick={[Function]}
    type="button"
  >
    <Icon
      inline={false}
      name="clipboard"
      passthroughProps={Object {}}
      size={18}
    />
  </button>
</Tooltip>
`;

exports[`CopyButton Defaults renders as expected 1`] = `
<Tooltip
  alignment="center"
  block={false}
  coloring="light"
  content="Copy"
  disabled={false}
  maxWidth="medium"
  padding="small"
  placement="top"
  respondsToClick={false}
  textSize="s"
>
  <button
    className="btn btn--xs py3 px3 round"
    data-clipboard-text="Copied by the default test case"
    onClick={[Function]}
    type="button"
  >
    <Icon
      inline={false}
      name="clipboard"
      passthroughProps={Object {}}
      size={18}
    />
  </button>
</Tooltip>
`;

exports[`CopyButton all props changes state on click 1`] = `
<Tooltip
  alignment="center"
  block={false}
  coloring="light"
  content="Copy"
  disabled={true}
  maxWidth="medium"
  padding="small"
  placement="top"
  respondsToClick={false}
  textSize="s"
>
  <button
    className="px6 py6 btn btn--purple btn--stroke"
    data-clipboard-text="more copiable text"
    data-test="copy-button"
    onClick={[Function]}
    type="button"
  >
    <Icon
      inline={false}
      name="check"
      passthroughProps={Object {}}
      size={18}
    />
    <Popover
      accessibleTitle={null}
      alignment="center"
      allowPlacementAxisChange={true}
      clickOutsideCloses={true}
      coloring="light"
      containWithinViewport={true}
      escapeCloses={true}
      getAnchorElement={[Function]}
      hasPointer={true}
      hideWhenAnchorIsOffscreen={true}
      padding="small"
      placement="top"
      receiveFocus={true}
      trapFocus={false}
      zIndex={1}
    >
      <div
        className="txt-s"
      >
        Copied!
      </div>
    </Popover>
  </button>
</Tooltip>
`;

exports[`CopyButton all props changes state on click 2`] = `
<Tooltip
  alignment="center"
  block={false}
  coloring="light"
  content="Copy"
  disabled={false}
  maxWidth="medium"
  padding="small"
  placement="top"
  respondsToClick={false}
  textSize="s"
>
  <button
    className="px6 py6 btn btn--purple btn--stroke"
    data-clipboard-text="more copiable text"
    data-test="copy-button"
    onClick={[Function]}
    type="button"
  >
    <Icon
      inline={false}
      name="clipboard"
      passthroughProps={Object {}}
      size={18}
    />
  </button>
</Tooltip>
`;

exports[`CopyButton all props renders as expected 1`] = `
<Tooltip
  alignment="center"
  block={false}
  coloring="light"
  content="Copy"
  disabled={false}
  maxWidth="medium"
  padding="small"
  placement="top"
  respondsToClick={false}
  textSize="s"
>
  <button
    className="px6 py6 btn btn--purple btn--stroke"
    data-clipboard-text="more copiable text"
    data-test="copy-button"
    onClick={[Function]}
    type="button"
  >
    <Icon
      inline={false}
      name="clipboard"
      passthroughProps={Object {}}
      size={18}
    />
  </button>
</Tooltip>
`;
